### PR TITLE
Save metrics

### PR DESCRIPTION
**Pull request by zcain117**
- Add skeleton for metrics regression test code
- Allow test_train_mnist, test_train_cifar, and test_train_imagenet to save metrics to file

Tested that the 3 files can write to GCS or to disk with this change.

### CODE REVIEW

**Comment by taylanbil** [comment]
Maybe we could put these changes in a separate file and similar to how we use `python -m torch_xla.distributed.xla_dist` for distributed training, we can do something like

`python -m torch_xla.test -- python /usr/share/torch-xla-0.5/pytorch/xla/test/test_train_imagenet.py     --fake_data`  to test?

**Comment by dlibenzi** [comment]
> Maybe we could put these changes in a separate file and similar to how we use `python -m torch_xla.distributed.xla_dist` for distributed training, we can do something like
> 
> `python -m torch_xla.test -- python /usr/share/torch-xla-0.5/pytorch/xla/test/test_train_imagenet.py --fake_data` to test?

I am changing the metrics save to support MP parallelism, and GCS exports.
Ideally, we would just the environment vars, and add maybe one import and one extra line at the end to flush the collected metrics to GCS.

**Comment by dlibenzi** [comment]
> > Maybe we could put these changes in a separate file and similar to how we use `python -m torch_xla.distributed.xla_dist` for distributed training, we can do something like
> > `python -m torch_xla.test -- python /usr/share/torch-xla-0.5/pytorch/xla/test/test_train_imagenet.py --fake_data` to test?
> 
> I am changing the metrics save to support MP parallelism, and GCS exports.
> Ideally, we would just the environment vars, and add maybe one import and one extra line at the end to flush the collected metrics to GCS.

Actually, the export to GCS could be done by the wrapper script.
So really no changes in the model.

**Comment by dlibenzi** [comment]
So here is how it might work IMHO.
We have a wrapper script which correctly sets up the environment variable `XLA_METRICS_FILE`:

https://github.com/pytorch/xla/pull/1293

And run the real test.
On exit, it parses the raw metrics data and extracts the wanted bits.
Note that the metrics data will be exported at every step.
Then will post to GCS eventually.
This script will also do the difference checks and fail the tests.
In this way we leave the model's main file slim, and we confine all the testing intelligence into the script (Python of course, not Shell).

**Comment by zcain117** [comment]
> So here is how it might work IMHO.
> We have a wrapper script which correctly sets up the environment variable `XLA_METRICS_FILE`:
> 
> #1293
> 
> And run the real test.
> On exit, it parses the raw metrics data and extracts the wanted bits.
> Note that the metrics data will be exported at every step.
> Then will post to GCS eventually.
> This script will also do the difference checks and fail the tests.
> In this way we leave the model's main file slim, and we confine all the testing intelligence into the script (Python of course, not Shell).

Sounds good, so you think the wrapper can call a subprocess and use a tmp file to hold the final metrics, then read from the tmp file in the parent process and make asserts and/or write metrics somewhere permanent?

It seems like a ton of data if we save metrics for every step. Most important is probably the final metrics right?

The wrapper should also carry forward any failure from the subprocess

**Comment by dlibenzi** [comment]
> > So here is how it might work IMHO.
> > We have a wrapper script which correctly sets up the environment variable `XLA_METRICS_FILE`:
> > #1293
> > And run the real test.
> > On exit, it parses the raw metrics data and extracts the wanted bits.
> > Note that the metrics data will be exported at every step.
> > Then will post to GCS eventually.
> > This script will also do the difference checks and fail the tests.
> > In this way we leave the model's main file slim, and we confine all the testing intelligence into the script (Python of course, not Shell).
> 
> Sounds good, so you think the wrapper can call a subprocess and use a tmp file to hold the final metrics, then read from the tmp file in the parent process and make asserts and/or write metrics somewhere permanent?
> 
> It seems like a ton of data if we save metrics for every step. Most important is probably the final metrics right?
> 
> The wrapper should also carry forward any failure from the subprocess

Let's say that a single step produces 4KB of metrics.
Let's pick resnet50, and batch 128, training for 80 EPOCHs on 8 cores.
An EPOCH is about 1200 steps.
So:

1200 * 80 * 4096 = 375MB

And usually we do not run (though we could) metrics collections for tests in which we are measuring accuracy.
The script will have to have all the logic of setting up the environment, running a test, comparing results, etc...

**Comment by zcain117** [comment]
Thanks @dlibenzi and @taylanbil for the wrapper idea. Much cleaner now in the runner files, PTAL.

Also updated the PR description

File: torch_xla/utils/gcsfs.py [hunk 1]
```diff
@@ -212,7 +212,7 @@ def remove(path):
   bucket.delete_blob(bpath)
 
 
-def write(path, content):
+def write(path, content, service_account_filename=None):
```
Comment (dlibenzi) [inline torch_xla/utils/gcsfs.py]: No please. Bring down global configuration like this is a PITA.
```Shell
export GOOGLE_APPLICATION_CREDENTIAL=...
```

File: torch_xla/debug/metrics_test_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import argparse
+import os
+import subprocess
+import sys
+
+import torch_xla.utils.gcsfs as gcsfs
+
+
+_CLOUD_STORAGE_PREFIX = 'gs://'
+_TMP_FILENAME = '/tmp/pytorch_tpu_metrics_tmp'
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_test_wrapper.py]: Do not bolt file names like that. Generate with Python API.
Also, in MP replication mode, many files (one per core) will be created.
Create a temp folder (using Python tempfile API), and let the process write there.
Then remove on exit.

File: torch_xla/debug/metrics_test_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser(
+      description='PyTorch on TPU distrubuted training',
+      epilog=('Usage example: metrics_test_wrapper.py '
+          '--golden_metrics_filename="gcs://bucket/file" '
+          '--metrics_output_filename="gcs://bucket/file" '
+          '--service_account_filename="path/to/file" -- python train.py'))
+  parser.add_argument('--golden_metrics_filename', type=str, default=None)
+  parser.add_argument('--metrics_output_filename', type=str, default=None)
+  parser.add_argument('--service_account_filename', type=str, default=None)
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_test_wrapper.py]: This is better be handled via environment.

File: torch_xla/debug/metrics_test_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  parser.add_argument(
+      'positional',
+      nargs='+',
+      type=str,
+      help='The python command to run.')
+  FLAGS = parser.parse_args()
+  if not (FLAGS.golden_metrics_filename or FLAGS.metrics_output_filename):
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_test_wrapper.py]: How things are going to work?
Will we have an outer wrapper script, or this is the `runner`?
If the latter, it should essentially take almost no argument besides the test command line, and the `test name`.
We can have a root GCS folder where all the per-test config is stored, which we append the test name, and inside there we expect to find certain file/folder structure (I know, GCS does not really have folders 😄 ).

File: torch_xla/utils/gcsfs.py (hunk 1 above)
Comment (zcain117) [reply torch_xla/utils/gcsfs.py]: I switched to use env var - we'll see if this works for the Google use case. We run the daily tests in one GCS project but I want to write metrics history to a pytorch-specific GCS project. The test infra does some read/writes to the team-wide GCS project but hopefully those are all done by the time the PyTorch-specific code (including this) kicks in

File: torch_xla/debug/metrics_test_wrapper.py (hunk 2 above)
Comment (zcain117) [reply torch_xla/debug/metrics_test_wrapper.py]: Good call, done. Added a flag for distributed so we know where to read - LMK if there's a better way

File: torch_xla/debug/metrics_test_wrapper.py (hunk 4 above)
Comment (zcain117) [reply torch_xla/debug/metrics_test_wrapper.py]: For the Google use case, we'll change our daily tests to call this wrapper and set these flags in that test config. I think it will be easier to debug issues if we have a code history of the config rather than a config file in GCS

Comment (dlibenzi) [reply torch_xla/debug/metrics_test_wrapper.py]: That is not a Google case.
That is a kind of config that usually is also global, and you do not change every API call.

Comment (zcain117) [reply torch_xla/debug/metrics_test_wrapper.py]: I'm saying that I'm imagining there will be an outer wrapper script. We will modify our current outer wrapper script to call `docker run python metrics_test_wrapper --metrics_output_filename="metrics.txt" --golden_metrics_filename="goldent.txt" --python test/test_train_mnist.py --num_epochs=1` instead of what it used to call: `docker run python test/test_train_mnist.py --num_epochs=1`. The outer script will take some action depending on return code

File: torch_xla/debug/metrics_test_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+_XLA_METRICS_FILE = 'XLA_METRICS_FILE'
+
+
+def _run_subprocess(cmd, distributed=False):
+  # Use _XLA_METRICS_FILE to pass the metrics report from the subprocess
+  # to this process. Note that distributed training jobs write 1 file per
+  # core.
+  tmp_dir = tempfile.mkdtemp()
+  _, tmp_path = tempfile.mkstemp(dir=tmp_dir)
+  file_to_read = '{}.0'.format(tmp_path) if distributed else tmp_path
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_test_wrapper.py]: I'd do a glob(), sort, and pick first.
In case of single core there will be no .0

File: torch_xla/debug/metrics_test_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  if not (FLAGS.golden_metrics_filename or FLAGS.metrics_output_filename):
+    raise ValueError('At least one of golden_metrics_filename or '
+                     'metrics_output_filename is required.')
+
+  metrics, sp_return_code = _run_subprocess(FLAGS.positional,
+                                            distributed=FLAGS.distributed)
+
+  # Include the params for this invocation when saving metrics.
+  output_string = '{}\n\n{}'.format(FLAGS, metrics)
+  if FLAGS.service_account_filename:
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_test_wrapper.py]: I thin this stuff should be configured externally, more then getting command line arguments and exporting variables.
The is part of the GCP configuration for a VM.

File: torch_xla/debug/metrics_test_wrapper.py (hunk 5 above)
Comment (zcain117) [reply torch_xla/debug/metrics_test_wrapper.py]: Good idea - switched to that and removed the --distributed flag

File: torch_xla/debug/metrics_test_wrapper.py (hunk 6 above)
Comment (zcain117) [reply torch_xla/debug/metrics_test_wrapper.py]: SGTM - I think I can make that work on the calling end. It's only tricky since the test runs in 1 GCP but my desired metrics location is in a different GCP. If I can't make it work I'll consider re-adding this flag

**Comment by zcain117** [comment]
> I would create a torch_xla/test folder and put this there.
> Move there even test/test_utils.py

Moved this to torch_xla/test.

I agree test_util.py should be there too but I'll do that in a follow-up PR since it'll involve changing several files

File: torch_xla/debug/metrics_test_wrapper.py (hunk 4 above)
Comment (dlibenzi) [reply torch_xla/debug/metrics_test_wrapper.py]: This needs more configuration to work, and we will end up having to have another "thick" wrapper script.
Better this be The Script, modulo eventually a trivial shell script that sets up some environment and run this.

For every test, beside the test/model data, this script will need some input configuration.
So what I am saying is lets have a root folder, say `gs://pytorch-tpu-test/` and lets call this script as:
```
test_runner.py --root=gs://pytorch-tpu-test/ --test_name=resnet50_blah -- CMD ARGS...
```

Then the test_runner will find, inside the `gs://pytorch-tpu-test/resnet50_blah/` folder all the info it needs.
For example, there will be a `gs://pytorch-tpu-test/resnet50_blah/golden/` folder to store the goldens.
There might also, for example, a `gs://pytorch-tpu-test/resnet50_blah/config.json` to store test specific config (tolerances, notify emails, ...).
You get the idea.

**Comment by zcain117** [comment]
Switched to the `--root` and `--test_folder_name` approach that we discussed

File: torch_xla/test/metrics_test_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  # 'base_metrics_file.0', 'base_metrics_file.1', etc.
+  metrics_files = glob.glob('{}*'.format(base_metrics_file))
+  if len(metrics_files) == 1:
+      # Non-distributed case: the correct file is simply 'base_metrics_file'.
+    file_to_read = base_metrics_file
+  else:
+    # Choose the file of the first core. When sorted alphabetically, the first
+    # element will be 'base_metrics_file' which is empty in the distributed
+    # case. The second element will be 'base_metrics_file.0'.
+    file_to_read = sorted(metrics_files)[1]
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: I think we should make this logic a bit more robust.
Like checking files size or something along those lines.

File: torch_xla/test/metrics_test_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  # Cleanup.
+  shutil.rmtree(tmp_dir)
+  del os.environ[_XLA_METRICS_FILE]
+
+  # Use only the metrics from the last step.
+  metrics = metrics[metrics.rfind('[MetricsData;'):]
+  return metrics, sp_return_code
+
+
+def _write_to_disk(output_string, output_filename):
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: How do we handle TTL to expire old files?
Manually?

File: torch_xla/test/metrics_test_wrapper.py (hunk 7 above)
Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: Sounds good, switched to that

File: torch_xla/test/metrics_test_wrapper.py (hunk 8 above)
Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: The plan was to avoid any kind of TTL for now since we want to hold onto our historical data and the storage cost is low. If we run into issues and need TTL, I can set it on the cloud storage bucket that holds the metrics history. No need for logic in this file

Comment (dlibenzi) [reply torch_xla/test/metrics_test_wrapper.py]: We will have to glob there, so having 1000s of file might be an issue.
But we could come back later to it, eventually.

Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: I see. We could maybe have a cronjob to move files from a `recent_history` directory to a `history` directory such that `recent_history`. I could probably have this as an automated workflow in the cloud project. But like you said, we can revisit when we have more files
